### ISSUE TITLE
Electron textarea can raise unexpected menu-related accessibility events when typing is in progress

### ISSUE THREAD

**Issue by guybark** [electron/electron#27826]
### Preflight Checklist
* [x ] I have read the [Contributing Guidelines](https://github.com/electron/electron/blob/master/CONTRIBUTING.md) for this project.
* [x ] I agree to follow the [Code of Conduct](https://github.com/electron/electron/blob/master/CODE_OF_CONDUCT.md) that this project adheres to.
* [x ] I have searched the issue tracker for an issue that matches the one I want to file, without success.

### Issue Details

* **Electron Version:**
  * 11.0.4
* **Operating System:**
  * Windows 10 Version 20H2 (OS Build 19042.804)
* **Last Known Working Electron version:**
  * Unknown.

### Expected Behavior
Typing into a textarea in an app should not lead to EVENT_SYSTEM_MENUSTART and EVENT_SYSTEM_MENUEND WinEvents events being raised by the app. 

### Actual Behavior
When typing into the textareas referenced in the “To Reproduce” section below, EVENT_SYSTEM_MENUSTART and EVENT_SYSTEM_MENUEND WinEvents are raised by the app whenever a character is added to the end of the content in the editable areas.  For more information on the related WinEvents, please visit [Event Constants (Winuser.h)](https://docs.microsoft.com/en-us/windows/win32/winauto/event-constants).

### To Reproduce
1. Move to an Electron textarea using either of the following repro steps.
•	Download and install GitHub desktop.
•	In GitHub Desktop, create a new repository so that you can get to the "Changes" view.
•	Verify that a "Description" textarea is shown.
OR
•	Download and install Visual Studio Code.
•	Open Help and select Report Issue.
•	Verify that a "Steps to reproduce" textarea is shown.

2. Run the AccEvent Windows SDK tool, as follows.
•	Run the AccEvent Windows SDK tool. If VS is already installed on the machine, the AccEvent.exe file may be in a folder such as “C:\Program Files (x86)\Windows Kits\10\bin\10.0.19041.0\x64”. More details on downloading the Windows 10 SDK are at [Windows 10 SDK - Windows app development](https://developer.microsoft.com/en-US/windows/downloads/windows-10-sdk/).
•	In the running AccEvent, select Mode, then “WinEvents (Out of Context)”.
•	Select Mode again, then “Settings”. In the Settings window’s Events list, leave only the items SYS_MENUSTART and SYS_MENUEND selected. Click OK.
•	Select Events, then “Start Listening”.

3. Move keyboard focus back to the textarea in the app being used to repro the issue.
4. Start typing into the textarea.

Result: The AccEvent tool reports that EVENT_SYSTEM_MENUSTART and EVENT_SYSTEM_MENUEND WinEvents are being raised in response to each character being typed at the end of the string in the textarea.

### Screenshots
The following screenshots shows the AccEvent Settings window being set up to report the menu-related WinEvents of interest, and the AccEvent tool subsequently reporting those events when typing into the GitHub Desktop app.

![image](https://user-images.githubusercontent.com/14315611/108564267-afdcd900-72b7-11eb-8702-8b60587c166e.png)

![image](https://user-images.githubusercontent.com/14315611/108564372-d26ef200-72b7-11eb-9042-65ae11cadaa7.png)

### Additional Information
This issue is being logged in response to an investigation into unexpected behavior when running the Windows Magnifier app while typing into the Electron textarea. If Magnifier is running at high magnification, then the magnified view can jump around while typing, leading to a very problematic experience for the person typing. If the top left corner of the editable area is outside of the magnified view, then with each key press the magnified view jumps to try to show both the updated position of the text insertion point in the editable area, and also the top left corner of the editable area, and both of those may not fit concurrently in the view. Typically the view is left at the top left corner of the editable area, meaning that customers are not being presented with the text they’ve just typed.

When investigating this issue, it seems that with each key press while typing, the Magnifier app is being told that the text is changing, which is expected, and also that keyboard focus is moving into the textarea. That focus changed event is not expected given that keyboard focus hasn’t moved while typing. One likely explanation is the focus changed event is being generated by Windows UI Automation in response to the Electron app raising the menu-related WinEvents. That is, the menu-related WinEvents give the impression that a menuitem has been selected while the typing is in progress, and so when typing continues in the editable area, it may mean that the keyboard focus has moved from that menuitem back into the editable area. The combination of focus apparently changing while the typing is in progress is leading to Windows Magnifier unexpectedly changing its view while typing. 

Given that no actual menu activity is occurring while the typing is in progress, it would seem appropriate for the Electron app to not raise the EVENT_SYSTEM_MENUSTART and EVENT_SYSTEM_MENUEND WinEvents. If those events are not raised, it may prevent the unexpected view changes when typing into the textarea while Windows Magnifier is being used.

**Comment by ckerr** [comment]
Hi @guybark, thanks for reaching out and reporting this, and in particular thanks for the attention to detail in the bug report.

As a first thought to help divide the problem space -- Maybe we can remove VSC / GitHub Desktop from the equation and try a simpler setup to triage?

What I have in mind, if you've still got your test setup handy:
* Instal [Electron Fiddle](https://www.electronjs.org/fiddle),  which is a app to facilitate running sample Electron code on arbitrary versions of Electron.
* Start Fiddle
* add a `<textarea id="test"></textarea>` to the `index.html`. (Alternately, load [this gist URL](https://gist.github.com/ckerr/720f5ecb2d298075d278956df7340bd7) via the gist entry in the upper right-hand corner of Fiddle's main window).
* Choose a version of Electron to test with using the selector in the upper-left of Fiddle's main window, e.g. Electron 11.0.4 as in the original issue description. Fiddle will download and unzip that version.
* Click the run button to run the code.
* A "Hello World!" app will appear with a textarea. Repeat your steps to trigger & monitor a11y events as you did with VSC / GitHub Desktop sessions.
* If this yields interesting results, you can also try testing with different versions of Electron (e.g. 10.0.0 or the latest 12.0.0-beta) using the same steps as above to see if the behavior changes.

Off the top of my head, I can't think of anything special that Electron is doing in its textareas and my first guess is that this may be happening inside the Chromium code. So I'd also be interested to know if the event series is equivalent when exercising a textarea inside Chromium? The "Hello World!" app includes version numbers. For example, if you run it with Electron 11.0.4 the app will read, `We are using Node.js 12.18.3, Chromium 87.0.4280.67, and Electron 11.0.4.` Thus you could download Chromium 87.0.4280.67 and measure the event sequences when using a textarea _in that version_ of Chromium, and comparing it to the sequence generated by Electron.

**Comment by guybark** [comment]
Hi @ckerr, thanks for all the helpful suggestions. 

I ran through your steps, and found I could reproduce the textarea in the Fiddle app apparently raising UIA FocusChanged events. This occurred with both Electron v11.2.3 and v11.0.4. This is the body of the test app:

```
<body>
  <div id="testlabel">Test</div>
  <textarea aria-labelledby="testlabel"></textarea> 
</body>
```


I said "apparently" above because I still think it's likely that the textarea itself isn't raising the UIA FocusChanged events. Rather, with each key press, the old EVENT_SYSTEM_MENUSTART and EVENT_SYSTEM_MENUEND WinEvents are being raised by something, and UIA is deciding in the belief that a menu is being opened and closed that it's appropriate for UIA itself to raise a UIA FocusChanged event to make sure that UIA clients like Magnifier know that focus is still in the textarea after the menu’s been opened and closed. If the EVENT_SYSTEM_MENUSTART and EVENT_SYSTEM_MENUEND events weren't generated on each key stroke, I think it's likely that the unexpected UIA FocusChanged events wouldn't be raised. I should say that I'm speculating here, but what I'm suggesting does fit the evidence I've seen so far.

It's worth adding that while I consistently find the UIA FocusChanged events being raised while typing in the textarea, the impact is not consistent. I've typed into the textarea with Magnifier running at a high zoom level, and sometimes the Magnifier experience is fine, and sometimes the magnified view bounces around unexpectedly while I type. I suspect that whether the UIA FocusChanged is problematic in practice here depends on the timing of the events raised. In one of my tests with the Fiddle app, I found that within a few 10s of milliseconds a text-related event was raised after the unexpected FocusChanged event was raised. If Magnifier has some threshold relating to when a FocusChanged event should be acted on, then it might have not got round to changing its view in response to the FocusChanged event before the text-related event came in and Magnifier acted on that event instead. (Often UIA clients won't act on every FocusChanged event, because when switching apps, focus can bounce around quickly before settling down somewhere.) But in another test with the Fiddler app, I found the text-related events arrived over 100 ms after the FocusChanged event. So in that case Magnifier might have begun changing its view to move the top left corner of the textarea into view in response to the FocusChanged event, and then quickly changed the view again in response to other events.

So for people trying to repro the Magnifier behavior, I think that'll be impacted by the timing of events on the particular device. But regardless of the Magnifier behavior, I do think it'd be interesting to explore how the menu-related events might be getting raised simply in response to a key press when typing a character into the textarea.

I've included an image below of the Accessibility Insights for Windows tool reporting the UIA events being raised as I type into the textarea in the Fiddle app. With each key press, a FocusChanged event is raised, followed by a number of other events, including 3 text-related ones.

Thanks again for your help.

![UIAEvents_textarea](https://user-images.githubusercontent.com/14315611/109727430-31125680-7b69-11eb-8bb0-47a32000301e.png)
